internal/protobufjs/ts_proto_library.bzl

```diff
@@ -0,0 +1,150 @@
+  js_tmpl_file = actions.declare_file(output_name + suffix + ".tmpl")
+
+  # Reference of arguments:
+  # https://github.com/dcodeIO/ProtoBuf.js/#pbjs-for-javascript
+  args = actions.args()
+  args.add(["--target", "static-module"])
+  args.add(["--wrap", wrap])
+  args.add("--strict-long") # Force usage of Long type with int64 fields
+  args.add(["--out", js_file.path + ".tmpl"])
+  args.add([f.path for f in proto_files])
+
+  actions.run(
+      executable = executable._pbjs,
+      inputs = proto_files,
+      outputs = [js_tmpl_file],
+      arguments = [args],
+  )
+
+  actions.expand_template(
+      template = js_tmpl_file,
+      output = js_file,
+      substitutions = {
+          # convert anonymous AMD module
+          #  define(["protobufjs/minimal"], function($protobuf) {
+          # to named
+          #  define("wksp/path/to/module", ["protobufjs/minimal"], ...
+          "define([": "define('%s/%s', [" % (amd_name, output_name),
+      }
+  )
+  return js_file
+
+def _run_pbts(actions, executable, js_file):
+  ts_file = actions.declare_file(js_file.basename[:-len(".closure.js")] + ".d.ts")
+
+  # Reference of arguments:
+  # https://github.com/dcodeIO/ProtoBuf.js/#pbts-for-typescript
+  args = actions.args()
+  args.add(["--out", ts_file.path])
+  args.add(js_file.path)
+
+  actions.run(
+      executable = executable._pbts,
+      progress_message = "Generating typings from %s" % js_file.short_path,
+      inputs = [js_file],
+      outputs = [ts_file],
+      arguments = [args],
+  )
+  return ts_file
+
+def _ts_proto_library(ctx):
+  sources = depset()
+  for dep in ctx.attr.deps:
+    if not hasattr(dep, "proto"):
+      fail("ts_proto_library dep %s must be a proto_library rule" % dep.label)
+    # TODO(alexeagle): go/new-proto-library suggests
+    # > should not parse .proto files. Instead, they should use the descriptor
+    # > set output from proto_library
+    # but protobuf.js doesn't seem to accept that bin format
+    sources = depset(transitive = [sources, dep.proto.transitive_sources])
+
+  output_name = ctx.attr.output_name or ctx.label.name
+
+  js_es5 = _run_pbjs(ctx.actions, ctx.executable, output_name, sources,
+                     amd_name = "/".join([p for p in [
+                         ctx.workspace_name,
+                         ctx.label.package,
+                     ] if p]))
+  js_es6 = _run_pbjs(ctx.actions, ctx.executable, output_name, sources,
+                     suffix = ".closure.js", wrap = "es6")
+  dts = _run_pbts(ctx.actions, ctx.executable, js_es6)
+
+  # Return a structure that is compatible with the deps[] of a ts_library.
+  return struct(
+    files = depset([dts]),
+    typescript = struct(
+      declarations = [dts],
+      transitive_declarations = [dts],
+      type_blacklisted_declarations = [],
+      es5_sources = depset([js_es5]),
+      es6_sources = depset([js_es6]),
+      transitive_es5_sources = depset(),
```

Comment:
I tried this PR out and had issues because the transitive sources here don't include the normal sources. I think transitive should be direct + indirect sources? In my case this causes issues when this read by [collect_es6_sources](https://github.com/bazelbuild/rules_nodejs/blob/befbd8961c18de98e254bea94db08ede78889f86/internal/common/collect_es6_sources.bzl#L36) which only reads `transitive_es6_sources`...

Reply:
Thanks! I didn't test it in a production bundle, that's missing here...